File: Problems/Algorithms/1026. Maximum Difference between Node and Ancestor/MaxDifferenceBetweenNodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxAncestorDiff(TreeNode root) {
        if (root == null) {
            return 0;
        }

        return maxAncestorDiff(root, root.val, root.val);
    }

    private int maxAncestorDiff(TreeNode root, int min, int max) {
        if (root == null) {
            return max - min;
        }

        min = Math.min(min, root.val);
        max = Math.max(max, root.val);

        return Math.max(maxAncestorDiff(root.left, min, max), maxAncestorDiff(root.right, min, max));
    }
}

File: Problems/Algorithms/1026. Maximum Difference between Node and Ancestor/max_difference_between_nodes.py
#!/usr/bin/env python3

from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        def helper(root, min_val, max_val):
            if not root:
                return max_val - min_val

            min_val = min(min_val, root.val)
            max_val = max(max_val, root.val)

            return max(helper(root.left, min_val, max_val), helper(root.right, min_val, max_val))

        if not root:
            return 0

        return helper(root, root.val, root.val)
